### ISSUE TITLE
TLS crash during establishing of the secure connection

### ISSUE THREAD

**Issue by ikoveshnikov** [tempesta-tech/tempesta#1187]
Just caught the bug on current master of Tempesta 5adf46dd9b94c9fa67bdd0233e3b836bc4931ad9 and the kernel tempesta-tech/linux-4.14.32-tfw@e90410720527a639ad6f50f0193e1ff00e7f0076.

First issue happen during build, same issue reproduced on CI:
```
...
  CC [M]  /home/user/tempesta/tempesta/tempesta_fw/t/unit/test_wq.o
  CC [M]  /home/user/tempesta/tempesta/tempesta_fw/t/unit/test_sched_hash.o
  CC [M]  /home/user/tempesta/tempesta/tempesta_fw/t/unit/test_tfw_str.o
  CC [M]  /home/user/tempesta/tempesta/tempesta_fw/t/unit/test_mem_fast.o
  CC [M]  /home/user/tempesta/tempesta/tempesta_fw/t/unit/test_hash.o
  CC [M]  /home/user/tempesta/tempesta/tempesta_fw/t/unit/test_sched_ratio.o
  LD [M]  /home/user/tempesta/tempesta/tempesta_fw/t/unit/tfw_test.o
  Building modules, stage 2.
  MODPOST 7 modules
WARNING: modpost: Found 1 section mismatch(es).
To see full details build your kernel with:
'make CONFIG_DEBUG_SECTION_MISMATCH=y'
  CC      /home/user/tempesta/tempesta/lib/tempesta_lib.mod.o
  CC      /home/user/tempesta/tempesta/tempesta_db/core/tempesta_db.mod.o
  CC      /home/user/tempesta/tempesta/tempesta_fw/t/tfw_bomber.mod.o
  CC      /home/user/tempesta/tempesta/tempesta_fw/t/tfw_fuzzer.mod.o
...
```

Then I've started the Tempesta with configuration as follows:
```
server 127.0.0.1:8080;

listen 443  proto=https;

tls_certificate /home/user/tls/tfw-root.crt;
tls_certificate_key /home/user/tls/tfw-root.key;
```

And used curl to access the site (uri `/hello` is text file length of 20 bytes):
```
% curl -v https://192.168.122.12/hello                                       :(
* Expire in 0 ms for 6 (transfer 0x55b06a492e80)
*   Trying 192.168.122.12...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x55b06a492e80)
* Connected to 192.168.122.12 (192.168.122.12) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
```

And i got the crash:
```
debian login: [  174.069883] [tdb] Start Tempesta DB
[  174.087312] [tempesta fw] Initializing Tempesta FW kernel module...
[  174.090478] [tempesta fw] Registering new classifier: frang
[  174.094256] [tempesta fw] Registering new scheduler: hash
[  174.099204] [tempesta fw] Registering new scheduler: ratio
[  174.147179] [tempesta fw] Preparing for the configuration processing.
[  174.150050] [tempesta fw] Configuration processing is completed.
[  174.165626] [tdb] Opened table /opt/tempesta/db/filter.tdb: size=16777216 rec_size=20 base=ffff9708b8c00000
[  174.341444] [tdb] Opened table /opt/tempesta/db/cache.tdb: size=268435456 rec_size=0 base=ffff9708a8c00000
[  174.344001] [tempesta fw] Open listen socket on: 0.0.0.0:443
[  174.360857] [tempesta fw] modules are started
[  304.429033] ------------[ cut here ]------------
[  304.430253] kernel BUG at /home/user/tempesta/tempesta/tls/tls_srv.c:706!
[  304.431633] invalid opcode: 0000 [#1] SMP PTI
[  304.432388] Modules linked in: tempesta_fw(O) tempesta_db(O) tempesta_tls(O) tempesta_lib(O) sha256_ssse3 sha512_ssse3 sha512_generic ccm ctr ghash_generic gf128mul gcm kvm_intel kvm irqbypass snd_hda_codec_generic crct10dif_pclmul iTCO_wdt crc32_pclmul iTCO_vendor_support snd_hda_intel snd_hda_codec snd_hda_core snd_hwdep evdev snd_pcm snd_timer qxl snd pcspkr serio_raw soundcore ttm lpc_ich mfd_core virtio_balloon sg virtio_console drm_kms_helper drm shpchp button binfmt_misc ip_tables x_tables autofs4 ext4 crc16 mbcache jbd2 crc32c_generic fscrypto ecb sr_mod cdrom crc32c_intel virtio_blk virtio_net ahci libahci aesni_intel aes_x86_64 psmouse crypto_simd i2c_i801 cryptd glue_helper ehci_pci libata sym53c8xx uhci_hcd scsi_transport_spi ehci_hcd scsi_mod usbcore usb_common virtio_pci virtio_ring
[  304.451094]  virtio [last unloaded: tempesta_lib]
[  304.453290] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G        W  O    4.14.0-tempesta-amd64 #1 Debian 4.14.32-tfw5-1
[  304.457386] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.12.0-20181126_142135-anatol 04/01/2014
[  304.460858] task: ffffffffb4c12480 task.stack: ffffffffb4c00000
[  304.462981] RIP: 0010:ttls_handshake_server_step+0x174e/0x1bc0 [tempesta_tls]
[  304.465666] RSP: 0018:ffff9708bfc03918 EFLAGS: 00010293
[  304.467570] RAX: 0000000000000047 RBX: ffff970895f82ab8 RCX: 0000000000000020
[  304.470188] RDX: 0000000000000020 RSI: ffff97089990d48c RDI: ffff970895f82ab8
[  304.472774] RBP: ffff97089990d48c R08: ffff9708bfc03a7c R09: 00000000000001c7
[  304.474614] R10: ffff970896b26ac0 R11: 0000000000000000 R12: 0000000000000003
[  304.476789] R13: 0000000000000000 R14: ffff9708bfc03a7c R15: ffff97089990d48c
[  304.479023] FS:  0000000000000000(0000) GS:ffff9708bfc00000(0000) knlGS:0000000000000000
[  304.481930] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  304.484056] CR2: 0000000002484008 CR3: 000000007ac0a002 CR4: 00000000003606f0
[  304.486644] Call Trace:
[  304.487570]  <IRQ>
[  304.488314]  ? sha512_update+0x1f/0x170 [sha512_ssse3]
[  304.490244]  ? crypto_sha512_update+0x4d/0x124 [sha512_generic]
[  304.492428]  ? crypto_shash_update+0x47/0x130
[  304.493999]  ? shash_async_import+0x40/0x40
[  304.495581]  ttls_recv+0x2d8/0x770 [tempesta_tls]
[  304.497132]  ? ttls_decrypt+0x570/0x570 [tempesta_tls]
[  304.498909]  ss_skb_process+0x19b/0x200 [tempesta_fw]
[  304.500934]  tfw_tls_msg_process+0xea/0x350 [tempesta_fw]
[  304.502794]  ? tfw_tls_msg_process+0x6f/0x350 [tempesta_fw]
[  304.506371]  ? __slab_free+0xa9/0x330
[  304.508117]  __gfsm_fsm_exec+0x56/0x90 [tempesta_fw]
[  304.510088]  tfw_connection_recv+0x4e/0x70 [tempesta_fw]
[  304.511935]  ? tfw_connection_send+0x30/0x30 [tempesta_fw]
[  304.513355]  ss_tcp_process_data+0x1db/0x440 [tempesta_fw]
[  304.514449]  ss_tcp_data_ready+0x43/0x90 [tempesta_fw]
[  304.516322]  tcp_data_queue+0x4f5/0xc50
[  304.517707]  tcp_rcv_established+0x27c/0x570
[  304.519252]  tcp_v4_do_rcv+0x129/0x1d0
[  304.520550]  tcp_v4_rcv+0x947/0xa50
[  304.521816]  ip_local_deliver_finish+0x9a/0x1c0
[  304.523466]  ip_local_deliver+0x6b/0xe0
[  304.524855]  ? tcp_v4_early_demux+0x112/0x150
[  304.526434]  ? ip_rcv_finish+0x17a/0x400
[  304.527854]  ip_rcv+0x289/0x3c0
[  304.529010]  ? inet_del_offload+0x40/0x40
[  304.530169]  __netif_receive_skb_core+0x84f/0xb30
[  304.531650]  ? process_backlog+0xa3/0x160
[  304.532939]  process_backlog+0xa3/0x160
[  304.535316]  net_rx_action+0x28e/0x3f0
[  304.537013]  ? __next_timer_interrupt+0x9e/0xc0
[  304.539052]  __do_softirq+0x10f/0x2a8
[  304.540489]  irq_exit+0xae/0xb0
[  304.541692]  call_function_single_interrupt+0x7d/0x90
[  304.543602]  </IRQ>
[  304.544387] RIP: 0010:native_safe_halt+0x2/0x10
[  304.546104] RSP: 0018:ffffffffb4c03e98 EFLAGS: 00000246 ORIG_RAX: ffffffffffffff04
[  304.548931] RAX: ffffffffb4297e60 RBX: ffffffffb4c12480 RCX: 0000000000000000
[  304.551630] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000000
[  304.554307] RBP: 0000000000000000 R08: 0000004f29248d5c R09: ffff9708bcc35b00
[  304.556981] R10: 0000000000000000 R11: 00000114197cedfb R12: ffffffffb4c12480
[  304.559619] R13: ffffffffb4c12480 R14: 0000000000000000 R15: 0000000000000000
[  304.561996]  ? __sched_text_end+0x3/0x3
[  304.563616]  default_idle+0x1a/0xf0
[  304.565399]  do_idle+0x16e/0x1f0
[  304.566812]  cpu_startup_entry+0x6f/0x80
[  304.568288]  start_kernel+0x462/0x482
[  304.569716]  secondary_startup_64+0xa5/0xb0
[  304.571307] Code: d5 ee ff ff 41 0f b6 04 24 49 83 c4 01 88 87 ad 04 00 00 e9 c5 f4 ff ff 4c 89 f8 41 bc 03 00 00 00 48 29 e8 89 c2 e9 ae ee ff ff <0f> 0b 49 89 eb 4d 89 f8 48 8b 6c 24 20 4d 89 ef 48 8b 73 10 4c 
[  304.578384] RIP: ttls_handshake_server_step+0x174e/0x1bc0 [tempesta_tls] RSP: ffff9708bfc03918

```

**Comment by ikoveshnikov** [comment]
Here is the crash with debug level 3:
```
[   86.584289] [tempesta fw]   client address: 192.168.122.1
[   86.586468] [tempesta fw]     client ffff8eb81bb063f0, conn_users=1
[   86.588160] [tempesta fw]       new client socket: sk=ffff8eb81a77a200, state=1
[   86.590737] [tempesta fw]     client ffff8eb81bb063f0, conn_users=2
[   86.592610] [tempesta fw]     tfw_http_conn_init: conn=[ffff8eb81bb920f8]
[   86.595846] [tempesta fw]       new client socket is accepted: sk=ffff8eb81a77a200, conn=ffff8eb81bb920f8, cli=ffff8eb81bb063f0
[   86.601060] [tempesta fw]       GFSM exec fsm 1, state 0x2200
[   86.603555] [tempesta tls]       ttls_recv: tls=ffff8eb81bb92420 len=62 read=0
[   86.605674] [tempesta tls]       input rec: type=22 ver=3:1 msglen=512 read=5 xfrm_ready=0
[   86.608499] [tempesta tls]   handshake message: msglen=512 type=1 hslen=508 read=9
[   86.611157] [tempesta tls]       handshake message 22 on state 0
[   86.612727] [tempesta tls]   server state: 0
[   86.614565] [tempesta tls]       enter TLS Server Handshake FSM at state 0x0(TTLS_CLIENT_HELLO)
[   86.617362] [tempesta tls]       enter TLS ClientHello FSM at state 0x0(TTLS_CH_HS_VER)
[   86.620699] [tempesta tls]       enter TLS ClientHello FSM at state 0x1(TTLS_CH_HS_RND)
[   86.623765] [tempesta tls]       dump 'ClientHello: random bytes ' (ptr=ffff8eb817c7daf8, 32 bytes):
[   86.627285] [tempesta tls]         00000000: 1b 35 dc a3 86 17 8c 83 05 21 80 d3 82 56 18 aa  .5.......!...V..
[   86.630875] [tempesta tls]         00000010: c3 c2 f9 fd 1d dd 38 7d c6 35 6c 7f 34 28 4a 92  ......8}.5l.4(J.
[   86.634573] [tempesta tls]       enter TLS ClientHello FSM at state 0x2(TTLS_CH_HS_SLEN)
[   86.637076] [tempesta tls]       ClientHello: Session ID length 32
[   86.639329] [tempesta tls]       enter TLS ClientHello FSM at state 0x3(TTLS_CH_HS_SESS)
[   86.642931] [tempesta tls]       Finish TLS ClientHello FSM at state 0x3, ret=-1
[   86.646010] [tempesta tls]     update checksum on buf ffff8eb8191e6487 len=57, hash=4
[   86.648673] [tempesta tls]       dump 'hash buf ' (ptr=ffff8eb8191e6487, 57 bytes):
[   86.651373] [tempesta tls]         00000000: 01 00 01 fc 03 03 1b 35 dc a3 86 17 8c 83 05 21  .......5.......!
[   86.655074] [tempesta tls]         00000010: 80 d3 82 56 18 aa c3 c2 f9 fd 1d dd 38 7d c6 35  ...V........8}.5
[   86.658744] [tempesta tls]         00000020: 6c 7f 34 28 4a 92 20 a2 5e 5e 1c c7 58 aa 9f 1f  l.4(J. .^^..X...
[   86.662084] [tempesta tls]         00000030: 2d 10 dd ea f8 c1 40 da f6                       -.....@..
[   86.665302] [tempesta tls]       ttls_recv: tls=ffff8eb81bb92420 len=455 read=62
[   86.668925] [tempesta tls]       handshake message 22 on state 3
[   86.671715] [tempesta tls]   server state: 3
[   86.673668] [tempesta tls]       enter TLS Server Handshake FSM at state 0x0(TTLS_CLIENT_HELLO)
[   86.676903] [tempesta tls]       enter TLS ClientHello FSM at state 0x3(TTLS_CH_HS_SESS)
[   86.679816] ------------[ cut here ]------------
[   86.680647] kernel BUG at /home/user/qtc/debug3/tempesta/tls/tls_srv.c:706!
[   86.682604] invalid opcode: 0000 [#1] SMP PTI
[   86.683867] Modules linked in: tempesta_fw(O) tempesta_db(O) sha256_ssse3 sha512_ssse3 sha512_generic ccm ctr ghash_generic gf128mul gcm tempesta_tls(O) tempesta_lib(O) binfmt_misc kvm_intel kvm irqbypass crct10dif_pclmul crc32_pclmul iTCO_wdt iTCO_vendor_support snd_hda_codec_generic qxl snd_hda_intel evdev ttm pcspkr serio_raw snd_hda_codec snd_hda_core snd_hwdep lpc_ich snd_pcm mfd_core snd_timer drm_kms_helper snd soundcore virtio_balloon sg virtio_console drm shpchp button ip_tables x_tables autofs4 ext4 crc16 mbcache jbd2 crc32c_generic fscrypto ecb sr_mod cdrom crc32c_intel virtio_blk virtio_net aesni_intel aes_x86_64 crypto_simd cryptd glue_helper ahci libahci psmouse libata i2c_i801 ehci_pci sym53c8xx uhci_hcd ehci_hcd scsi_transport_spi scsi_mod usbcore usb_common virtio_pci virtio_ring
[   86.706628]  virtio
[   86.707409] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G        W  O    4.14.0-tempesta-amd64 #1 Debian 4.14.32-tfw5-1
[   86.710818] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.12.0-20181126_142135-anatol 04/01/2014
[   86.714546] task: ffffffff9b612480 task.stack: ffffffff9b600000
[   86.716801] RIP: 0010:ttls_parse_client_hello+0x196d/0x2040 [tempesta_tls]
[   86.719367] RSP: 0018:ffff8eb83fc038b8 EFLAGS: 00010206
[   86.721362] RAX: 0000000000000047 RBX: 00000000000001c7 RCX: 0000000000000020
[   86.723989] RDX: 0000000000000020 RSI: 00000000000000f6 RDI: 000000000000083f
[   86.726618] RBP: 0000000000000003 R08: 00000000000a5427 R09: 000000000000036e
[   86.728981] R10: ffff8eb8191e64c0 R11: ffffffff9c38ae6d R12: 0000000000000003
[   86.730968] R13: ffff8eb83fc03a6c R14: ffff8eb81bb92420 R15: ffff8eb81bbd608c
[   86.734631] FS:  0000000000000000(0000) GS:ffff8eb83fc00000(0000) knlGS:0000000000000000
[   86.738207] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   86.740305] CR2: 000055982f2e1090 CR3: 0000000019e0a004 CR4: 00000000003606f0
[   86.742800] Call Trace:
[   86.743713]  <IRQ>
[   86.744322]  ttls_handshake_server_step+0x2a4/0xb20 [tempesta_tls]
[   86.746465]  ? ttls_update_checksum+0xf6/0x220 [tempesta_tls]
[   86.748563]  ttls_recv+0x325/0xa20 [tempesta_tls]
[   86.750270]  ? ttls_decrypt+0xbb0/0xbb0 [tempesta_tls]
[   86.752105]  ss_skb_process+0x19b/0x200 [tempesta_fw]
[   86.753816]  tfw_tls_msg_process+0xea/0x370 [tempesta_fw]
[   86.755754]  ? tfw_tls_msg_process+0x6f/0x370 [tempesta_fw]
[   86.758148]  __gfsm_fsm_exec+0x7e/0xc0 [tempesta_fw]
[   86.759784]  tfw_connection_recv+0x4e/0x70 [tempesta_fw]
[   86.762501]  ? tfw_connection_send+0x30/0x30 [tempesta_fw]
[   86.764652]  ss_tcp_process_data+0x1db/0x440 [tempesta_fw]
[   86.766626]  ss_tcp_data_ready+0x43/0x90 [tempesta_fw]
[   86.768459]  tcp_data_queue+0x4f5/0xc50
[   86.769779]  tcp_rcv_established+0x27c/0x570
[   86.771296]  tcp_v4_do_rcv+0x129/0x1d0
[   86.772633]  tcp_v4_rcv+0x947/0xa50
[   86.773911]  ip_local_deliver_finish+0x9a/0x1c0
[   86.775537]  ip_local_deliver+0x6b/0xe0
[   86.776921]  ? tcp_v4_early_demux+0x112/0x150
[   86.778481]  ? ip_rcv_finish+0x17a/0x400
[   86.779856]  ip_rcv+0x289/0x3c0
[   86.780926]  ? inet_del_offload+0x40/0x40
[   86.782448]  __netif_receive_skb_core+0x84f/0xb30
[   86.784325]  ? process_backlog+0xa3/0x160
[   86.785849]  process_backlog+0xa3/0x160
[   86.787319]  net_rx_action+0x28e/0x3f0
[   86.788741]  ? __next_timer_interrupt+0x9e/0xc0
[   86.790831]  __do_softirq+0x10f/0x2a8
[   86.791958]  irq_exit+0xae/0xb0
[   86.793775]  call_function_single_interrupt+0x7d/0x90
[   86.795738]  </IRQ>
[   86.796336] RIP: 0010:native_safe_halt+0x2/0x10
[   86.797705] RSP: 0018:ffffffff9b603e98 EFLAGS: 00000246 ORIG_RAX: ffffffffffffff04
[   86.798946] RAX: ffffffff9ac97e60 RBX: ffffffff9b612480 RCX: 0000000000000000
[   86.800406] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000000
[   86.802480] RBP: 0000000000000000 R08: 0000001790962066 R09: ffff8eb8178a8800
[   86.804622] R10: 0000000000000000 R11: 000000e1698e8cfb R12: ffffffff9b612480
[   86.806755] R13: ffffffff9b612480 R14: 0000000000000000 R15: 0000000000000000
[   86.808905]  ? __sched_text_end+0x3/0x3
[   86.810054]  default_idle+0x1a/0xf0
[   86.811177]  do_idle+0x16e/0x1f0
[   86.812335]  cpu_startup_entry+0x6f/0x80
[   86.813749]  start_kernel+0x462/0x482
[   86.815062]  secondary_startup_64+0xa5/0xb0
[   86.816453] Code: bc ff ff ff ff 48 89 44 24 20 89 c3 41 89 c7 e9 a7 ea ff ff 0f 0b 48 c7 c7 b0 76 68 c0 31 c0 e8 25 cd 06 da 0f 0b e9 1c fa ff ff <0f> 0b 48 8b 5c 24 30 e9 7c f4 ff ff 0f 0b 4c 2b 7c 24 08 bd 03 
[   86.824040] RIP: ttls_parse_client_hello+0x196d/0x2040 [tempesta_tls] RSP: ffff8eb83fc038b8
```

**Comment by ikoveshnikov** [comment]
Just noticed, I have the following issues in the system log:

This message appears when the Tempesta is started:
```
debian login: [   27.957522] tempesta_lib: loading out-of-tree module taints kernel.
[   27.978413] ghash_clmulni_intel: Unknown symbol cryptd_free_ahash (err -2)
[   27.981514] ghash_clmulni_intel: Unknown symbol cryptd_alloc_ahash (err -2)
[   27.982951] ghash_clmulni_intel: Unknown symbol cryptd_ahash_child (err -2)
[   27.984994] ghash_clmulni_intel: Unknown symbol cryptd_shash_desc (err -2)
[   27.986453] ghash_clmulni_intel: Unknown symbol cryptd_ahash_queued (err -2)
[   28.049424] [tdb] Start Tempesta DB
```

**Comment by krizhanovsky** [comment]
Maybe we need to add some crypt dipendency for Tempesta in Kconfig? It looks like you have hash routines compiled as modules, while there're compiled in in my VM. I'm attaching my kernel config just in case. 
[kern-config.txt](https://github.com/tempesta-tech/tempesta/files/2862387/kern-config.txt)

**Comment by krizhanovsky** [comment]
@i-rinat reported also unit test issues, seems caused by my recent changes in `__init` specifier.

```
WARNING: /root/tempesta/tempesta_fw/t/unit/tfw_test.o(.text+0xfea5): Section mismatch in reference from the function test_suite__tls() to the function .init.text:ttls_mpi_modinit()
The function test_suite__tls() references
the function __init ttls_mpi_modinit().
This is often because test_suite__tls lacks a __init
annotation or the annotation of ttls_mpi_modinit is wrong.
```

```
./run_tests.py -v normalization

[  607.284147] ------------[ cut here ]------------
[  607.285077] WARNING: CPU: 0 PID: 9145 at kernel/module.c:1133 module_put+0x7a/0x90
[  607.285077] Modules linked in: tempesta_tls(O+) tempesta_lib(O) sha256_ssse3 sha512_ssse3 sha512_generic 9p fscache 9pnet crc32_pclmul crc32c_intel ghash_clmulni_intel pcbc aesni_intel crypto_simd aes_x86_64 cryptd glue_helper virtio_balloon virtio_console virtio_net virtio_pci virtio_ring virtio [last unloaded: tempesta_lib]
[  607.285077] CPU: 0 PID: 9145 Comm: insmod Tainted: G        W  O    4.14.32-tfw #0.0.0
[  607.285077] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.12.0-1 04/01/2014
[  607.285077] task: ffff91d65a6c5580 task.stack: ffffae4d40730000
[  607.285077] RIP: 0010:module_put+0x7a/0x90
[  607.285077] RSP: 0018:ffffae4d40733c20 EFLAGS: 00010297
[  607.285077] RAX: 0000000000000001 RBX: ffff91d65a324448 RCX: 00000000ffffffff
[  607.285077] RDX: 0000000000000000 RSI: ffff91d67fc23c00 RDI: ffffffffc008c140
[  607.285077] RBP: ffff91d65abf0b80 R08: 0000000000023c00 R09: ffffffff8ad0c0a3
[  607.285077] R10: ffffd81284691d00 R11: 0000000000000001 R12: ffffffffc008c140
[  607.285077] R13: ffffffffc0147260 R14: ffffffffc0138ed8 R15: ffff91d65ab86360
[  607.285077] FS:  00007f349621e700(0000) GS:ffff91d67fc00000(0000) knlGS:0000000000000000
[  607.285077] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  607.285077] CR2: 00007f3495dd4e50 CR3: 0000000100380004 CR4: 00000000003606f0
[  607.285077] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  607.285077] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  607.285077] Call Trace:
[  607.285077]  crypto_destroy_tfm+0x5c/0xb0
[  607.285077]  crypto_destroy_tfm+0x33/0xb0
[  607.285077]  crypto_destroy_tfm+0x33/0xb0
[  607.285077]  ttls_crypto_modinit+0x83/0x16e [tempesta_tls]
[  607.285077]  ttls_init+0x1b/0xac [tempesta_tls]
[  607.285077]  ? 0xffffffffc0167000
[  607.285077]  do_one_initcall+0x3c/0x150
[  607.285077]  ? kmem_cache_alloc_trace+0x2c/0x150
[  607.285077]  do_init_module+0x56/0x1e5
[  607.285077]  load_module+0x146a/0x1b60
[  607.285077]  ? kernel_read_file+0x154/0x180
[  607.285077]  ? SYSC_finit_module+0xa2/0xb0
[  607.285077]  SYSC_finit_module+0xa2/0xb0
[  607.285077]  do_syscall_64+0x5b/0x100
[  607.285077]  entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[  607.285077] RIP: 0033:0x7f3495d4f229
[  607.285077] RSP: 002b:00007ffeb336ca08 EFLAGS: 00000246 ORIG_RAX: 0000000000000139
[  607.285077] RAX: ffffffffffffffda RBX: 00005566f3056220 RCX: 00007f3495d4f229
[  607.285077] RDX: 0000000000000000 RSI: 00005566f2a5a638 RDI: 0000000000000004
[  607.285077] RBP: 00005566f2a5a638 R08: 0000000000000000 R09: 00007f3496006ea0
[  607.285077] R10: 0000000000000004 R11: 0000000000000246 R12: 0000000000000000
[  607.285077] R13: 00005566f30561f0 R14: 0000000000000000 R15: 0000000000000000
[  607.285077] Code: 85 db 74 21 48 8b 03 48 89 fd 48 8b 7b 08 48 83 c3 18 4c 89 e2 48 89 ee e8 e4 3e b3 00 48 8b 03 48 85 c0 75 e5 5b 5d 41 5c eb b8 <0f> 0b eb a6 89 c2 eb 8d 0f 1f 40 00 66 2e 0f 1f 84 00 00 00 00 
```

Maybe it makes sense to declare the test routines as `__init`?

**Comment by i-rinat** [comment]
The immediate issue with triggered BUG_ON in ClientHello parser was resolved in #1197. Section mismatch issues — in #1190. Warning triggered in `module_put()` was caused the same reasons as #1188, and was fixed there. Also, #1220 and #1231 were somehow related too.
